jobs: Ensure schedules are cancelled when scheduler disabled.

miretskiy opened the pull request:
Ensure that currently executing schedules are cancelled immediately
when jobs scheduler disabled via the `jobs.scheduler.enabled` setting.

Fixes #77248

Release Note (enterprise change): Currently executing schedules are
cancelled immediately when jobs scheduler disabled.

Release Justification: stability improvement.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77306)
<!-- Reviewable:end -->

Commit:
jobs: Ensure schedules are cancelled when scheduler disabled.

Ensure that currently executing schedules are cancelled immediately
when jobs scheduler disabled via the `jobs.scheduler.enabled` setting.

Fixes #77248

Release Note (enterprise change): Currently executing schedules are
cancelled immediately when jobs scheduler disabled.

Release Justification: stability improvement.

dt commented on pkg/jobs/job_scheduler.go:
```diff
@@ -375,6 +376,56 @@ func (s *jobScheduler) schedulerEnabledOnThisNode(ctx context.Context) bool {
+type syncCancelFunc struct {
+	syncutil.Mutex
+	context.CancelFunc
+	sv *settings.Values
```
do we need to capture sv? don't we have it at the withCancelOnDisabled callsite too?

miretskiy replied:
yeah; we could pass it in...done.

miretskiy commented:
Informs #70920

miretskiy commented:
bors r+
